docs/architecture/airbyte-specification.md

```diff
@@ -0,0 +1,180 @@
+
+#### Spec
+* Input:
+    1. none.
+* Output:
+    1. spec - a [ConnectorSpecification](https://github.com/airbytehq/airbyte/blob/master/airbyte-protocol/models/src/main/resources/airbyte_protocol/airbyte_message.yaml#L133-L149).
+* A `ConnectorSpecification` contains information about the source.
+* The `connectionSpecification` of that struct must be valid JsonSchema. It describes what inputs are needed in order for the source to interact with the underlying data source. e.g. If using a postgres source, the `ConnectorSpecification` would specify that a `hostname`, `port`, and `password` are required in order for the integration to function. The UI reads the JsonSchema in this field in order to render the input fields for a user to fill in. This JsonSchema is also used to validate that the provided inputs are valid. e.g. If `port` is one of the fields and the JsonSchema in the `connectorSpecification` specifies that this filed should be a number, if a user inputs "airbyte", they will receive an error. Airbyte adheres to JsonSchema validation rules.
+
+#### Check
+* Input:
+    1. config: A configuration json object that has been validated using the `ConnectorSpecification`.
+* Output:
+    1. connectionStatus - an [AirbyteConnectionStatus](https://github.com/airbytehq/airbyte/blob/master/airbyte-protocol/models/src/main/resources/airbyte_protocol/airbyte_message.yaml#L94-L107).
+* It attempts to connect to the underlying data source in order to verify that the provided credentials are usable. e.g. If the given the credentials, it can connect to the postgres database, it will return a success response. If it fails (perhaps the password is incorrect), it will return a failed response and (when possible) a helpful error message.
+
+#### Discover
+* Input:
+    1. config: A configuration json object that has been validated using the `ConnectorSpecification`.
+* Output:
+    1. catalog - an [AirbyteCatalog](https://github.com/airbytehq/airbyte/blob/master/airbyte-protocol/models/src/main/resources/airbyte_protocol/airbyte_message.yaml#L108-L132).
+* This command detects the _structure_ of the data in the data source.
+* An `AirbyteCatalog` describes the structure of data in a data source. It has a single field called `streams` that contains a list of `AirbyteStream`s. Each of these contain a `name` and `json_schema` field. The `json_schema` field accepts any valid JsonSchema and describes the structure of a stream. This data model is intentionally flexible. That can make it a little hard at first to mentally map onto your own data, so we provide some example below:
+    * If you are using a data source that is a "traditional" relational database, each table in that database would make to an `AirbyteStream`. Each column in the table would be a key in the `json_schema` field.
+        * e.g. If we have a table called `users` which had the columns `name` and `age` (the age column is optional) the `AirbyteCatalog` would look like this:
```

Comment:
if the age column is optional, the would be `["null", "string"]` right?

Reply:
i believe both are valid JsonSchema.